File: python/algo/202511/algo2593.py
from typing import List

#滑动窗口与双指针 - 六、分组循环
class Solution:
    def findScore(self, nums: List[int]) -> int:
        # 自行解答
        ans, n = 0, len(nums)
        marked = [0] * n
        
        # 用 python 比 c++ 容易实现多了
        # 两种写法
        # for i, v in sorted(enumerate(nums), key=lambda p:p[1]):
        for v, i in sorted(zip(nums, range(n))):
            if (marked[i]): continue
            ans += v
            marked[i] = 1
            if i: marked[i-1] = 1
            if i+1 < n: marked[i+1] = 1
        return ans

File: cplus/leetcode/202511/algo2593.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 六、分组循环
class Solution {
public:
    long long findScore(vector<int>& nums) {
        // 自行解答，关键 stable_sort 要用稳定排序
        long long ans = 0;
        unordered_set<int> marked;
        int n = nums.size(), ids[n];
        iota(ids, ids+n, 0);
        stable_sort(ids, ids+n, [&](int i, int j){   // 稳定排序
            return nums[i] < nums[j];
        });         

        for (int i=0; i<n; i++) { 
            int inx = ids[i];
            if (marked.count(inx)) continue;
            ans += nums[inx];
            marked.insert(inx+1), marked.insert(inx), marked.insert(inx-1);
        }
        return ans;
    }

    long long findScore2(vector<int>& nums) {
        // 分组循环，按递减分段，每段最小值一定可以入选，然后倒着间隔 1 枚举
        // 用自己的分组循环模板，比题解的分组循环写更容易理解
        long long ans = 0;
        int i = 0, n = nums.size();
        while (i < n) {
            int st = i ++;
            while (i < n && nums[i-1]>nums[i]) i++;
            for (int j=i-1; j>=st; j-=2) ans += nums[j];
            i++; // i-1 选择 i 不选
        }
        return ans;
    }
};